flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -124,8 +131,15 @@ public PulsarWriter(
             throw new FlinkRuntimeException("Cannot initialize schema.", e);
         }
 
+        // Initialize metrics
+        this.pulsarSinkWriterMetrics = new PulsarSinkWriterMetrics(initContext.metricGroup());
+        this.lastSendTime = PulsarSinkWriterMetrics.INVALID_LAST_SEND_TIME;
         // Create this producer register after opening serialization schema!
-        this.producerRegister = new TopicProducerRegister(sinkConfiguration);
+        this.producerRegister =
+                new TopicProducerRegister(sinkConfiguration, pulsarSinkWriterMetrics);
```

Comment:
Consider keep pulsasrSinkWriterMetrics outside of topicProducerRegister, instead, all metrics registering code should happen in pulsarWriter.